src/Functions/array/arrayAggregation.cpp

```diff
@@ -152,9 +152,77 @@ struct ArrayAggregateImpl
         return result;
     }
 
+    template <AggregateOperation op = aggregate_operation>
+    requires(op == AggregateOperation::min || op == AggregateOperation::max)
+    static bool executeMinOrMax(const ColumnPtr & mapped, const ColumnArray::Offsets & offsets, ColumnPtr & res_ptr)
+    {
+        const ColumnConst * const_column = checkAndGetColumn<ColumnConst>(&*mapped);
+        if (const_column)
+        {
+            MutableColumnPtr res_column = const_column->getDataColumn().cloneEmpty();
+            res_column->insertMany(const_column->getField(), offsets.size());
+            res_ptr = std::move(res_column);
+            return true;
+        }
+
+        MutableColumnPtr res_column = mapped->cloneEmpty();
+        static constexpr int nan_null_direction_hint = aggregate_operation == AggregateOperation::min ? 1 : -1;
+
+        if (offsets.size() == 1 && offsets[0] != 0)
+        {
+            /// Fast path: Use permutation (which in some cases is optimized) to find the extreme value
```

Comment:
will we still have o(arr_size) in cases where it is not optimised?

Reply:
Yes, but that was still the case (it didn't get worse).

I could implement a similar approach as `executeType` done for the other functions but that a) requires declaring all optimizable types and a default implementation (such as this) with b) many templates and compilation time and c) I'd rather implement it in the permutation itself (accept begin and end) as it'd help improving many other cases, like min/max/argMin/argMax, etc where we have the same issue (for sparse columns for example).

Maybe it makes sense to remove the fast path for now, but it also helps as a reminder of a possible optimization.

Reply:
i don't quite understand how. get permutation should be nlogn in general case.

Reply:
Yes, but it can be o(n) when limit is 1. And it can be implemented with avx2:

https://github.com/ClickHouse/ClickHouse/blob/c15699e32d1168f292eca290ed45ab1eaa137da9/src/Columns/ColumnVector.cpp#L227-L246

Maybe I should just introduce a `getMinIndex(start, end)` for columns and use that instead (in another PR). I'll remove the code for now and leave it permutation-less for now